### PR TITLE
Add Email Service and automatic emails for company applications

### LINKED ISSUE #54: Company Applications and Approvals
- [x] Create CompanyApplication model: Has attributes of Account + Company (for creating them later on if the Application is accepted) + `approvalDate` + `submitDate` + `rejectDate` (the dates  - except approval - are null until changed to represent that that event did not yet happen)
- [x] List CompanyApplications endpoint (admin only)
- [x] CompanyApplication approval endpoint (admin only) -> Creates the Company and, in case of success, the Account for the Company representative
- [x] CompanyApplication rejection endpoint (admin only)

### PR DESCRIPTION
**Pull request by imnotteixeira**
Closes #54 

Added a generic Email Service that uses nodemailer to send emails as NIAEFEUP/NIJobs team.

Added the following automatic emails:
* Notify NIAEFEUP when a new application is created (sent to the MAIL_ADDRESS specified in .env file)
* Notify submission of application to the company email
* Notify approval of company to the company email

If no email is specified in the .env file, no emails will be sent and the backend won't crash

### CODE REVIEW

File: src/lib/nodemailer.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                refreshToken,
+                accessToken,
+            },
+            connectionTimeout: 30000
+        });
+    }
+
+    verifyConnection() {
+        if (!this.transporter) throw new Error("Trying to use Email Service without having a transporter set. Call init() first.");
+        return promisify(this.transporter.verify)();
```
Comment (miguelpduarte) [inline src/lib/nodemailer.js:28]: Afaik `nodemailer` already does this internally?
> If `callback` argument is not set then the method returns a Promise object. Nodemailer itself does not use Promises internally but it wraps the return into a Promise for convenience.

In https://nodemailer.com/usage/#sending-mail

This is not extremely explicit in the "docs" though... They seem to be quite lacking...

File: src/lib/nodemailer.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        if (!message.attachments || !(message.attachments instanceof Array)) {
+            message.attachments = [];
+            message.attachments.push({
+                filename: "logo-niaefeup.png",
+                path: "https://ni.fe.up.pt/images/logo-niaefeup.png",
+                cid: "id_1234698",
+            });
+        }
+
+        message.html += EMAIL_SIGNATURE;
```
Comment (miguelpduarte) [inline src/lib/nodemailer.js:61]: Should we consider use some sort of email templating to make this easier (and possibly safer)? Such as mustache/handlebars/etc

Comment (miguelpduarte) [reply src/lib/nodemailer.js:61]: We could have base templates that would use a layout that would have the signature and everything necessary, and others would just "extend" these.

File: src/services/emails/companyApplicationApproval.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+/* eslint-disable max-len */
+
```
Comment (miguelpduarte) [inline src/services/emails/companyApplicationApproval.js:2]: Yeah, the string interpolation in the html makes me kind of wary of template injection. We should definitely use some templating library that handles this before shipping this. Maybe this might be in a separate PR, but before having this go live with any sort of email-sending capabilities.

The templating would also be easier to grok than these strings, most likely.

Comment (miguelpduarte) [reply src/services/emails/companyApplicationApproval.js:2]: Someone in the board should also have a look at the emails, since they will be sent to companies. I think they are fine as-is, but they may benefit from some of their input or at least approval.

File: src/services/application.js [hunk 4]
```diff
@@ -2,6 +2,13 @@ const CompanyApplication = require("../models/CompanyApplication");
 const CompanyApplicationRules = require("../models/CompanyApplication").CompanyApplicationRules;
 const hash = require("../lib/passwordHashing");
 const AccountService = require("./account");
+const EmailService = require("../lib/nodemailer");
+const {
+    NEW_COMPANY_APPLICATION_ADMINS,
+    NEW_COMPANY_APPLICATION_COMPANY,
+    APPROVAL_NOTIFICATION,
+} = require("./emails/companyApplicationApproval");
```
Comment (miguelpduarte) [inline src/services/application.js:10]: No rejection email?

File: src/lib/nodemailer.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
+const promisify = require("util").promisify;
+const nodemailer = require("nodemailer");
+const EMAIL_SIGNATURE = require("../services/emails/signature");
+
```
Comment (miguelpduarte) [inline src/lib/nodemailer.js:4]: The name of the file should maybe be switched to `email.js` or `email_service.js` or something similar.

Having the name of the library as the file name might introduce some confusion as to whether the service or nodemailer itself is being used directly when browsing the code.

File: test/end-to-end/application.js [hunk 6]
```diff
@@ -84,6 +88,39 @@ describe("Company application endpoint test", () => {
...
+                    html: adminEmailOptions.html,
+                    to: config.mail_address
+                });
+
+                expect(EmailService.sendMail).toHaveBeenNthCalledWith(2, {
+                    subject: companyEmailOptions.subject,
+                    text: companyEmailOptions.text,
+                    html: companyEmailOptions.html,
+                    to: application.email
+                });
```
Comment (miguelpduarte) [inline test/end-to-end/application.js:121]: Can these fail if they ever occur in a different order? I.e. is there a way to make this check regardless of the order the method is called in?

**Review by miguelpduarte — CHANGES_REQUESTED** [review]
Overall looks good! Just have some suggestions!

Btw, the EmailService unit tests were quite tasty, props! :sunglasses:

File: src/services/application.js (hunk 4 above)
Comment (imnotteixeira) [reply src/services/application.js:10]: I thought about it, but I didn't know what to put there. "You're not good enough" sounded too bad :P

@andrefmrocha could use your input here

File: src/lib/nodemailer.js (hunk 2 above)
Comment (imnotteixeira) [reply src/lib/nodemailer.js:61]: https://github.com/yads/nodemailer-express-handlebars this might work

File: src/services/application.js (hunk 4 above)
Comment (andrefmrocha) [reply src/services/application.js:10]: What use cases are we thinking for rejecting a company application?

Comment (miguelpduarte) [reply src/services/application.js:10]: We could use the reject reason as some sort of message, besides the generic template.

Nevertheless, the notification should still exist and thus the email as well, regardless of not knowing what to say 😛

Comment (miguelpduarte) [reply src/services/application.js:10]: @andrefmrocha I've discussed this with you over PM but the idea is that a company is rejected if the admins don't like their application. For example their motivation is not valid, handling spam that might find the website and apply just because, etc.

Comment (andrefmrocha) [reply src/services/application.js:10]: It would seem that it would be tricky to create a generic email to reject a company since we have so many use cases. But probably an email saying they were rejected and forwarding any complaints to NI's email could be interesting. Wdyt @andrefmrocha @imnotteixeira ?

File: .env [hunk 7]
```diff
@@ -28,3 +28,22 @@ ADMIN_EMAIL=ni@aefeup.pt
 ADMIN_PASSWORD=n1j0bs_ftw.12345
 #CORS allowed origin - OVERRIDE IN PRODUCTION
 ACCESS_CONTROL_ALLOW_ORIGIN= 
+
+# Mail service information. If you don't provide a MAIL_ADDRESS, no emails will be sent. The app will execute no-ops and won't crash
+# However, if you want to send emails, you need to fill all of the following 5 fields
+# Check this SO Post for details on how to configure your personal account for testing (https://stackoverflow.com/questions/51933601/what-is-the-definitive-way-to-use-gmail-with-oauth-and-nodemailer)
+
+# The email address from which the emails are sent
+MAIL_ADDRESS=
```
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply .env:37]: Sure

File: .env [hunk 8]
```diff
@@ -28,3 +28,22 @@ ADMIN_EMAIL=ni@aefeup.pt
...
+MAIL_CLIENT_ID=
+
+# Client Secret used for OAuth
+MAIL_CLIENT_SECRET=
+
+# OAuth refresh token (for oauth token auto-refreshing)
+MAIL_REFRESH_TOKEN=
+
+#Initial OAuth access token, if expired, it will refresh using REFRESH_TOKEN
+MAIL_ACCESS_TOKEN=
```
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply .env:49]: Confirmed that it doesn't work in our prod email. Going back to simple SMTP only

File: package.json [hunk 9]
```diff
@@ -8,8 +8,8 @@
     "prod": "NODE_ENV=production node src/index.js",
     "lint": "eslint src test --max-warnings 0",
     "lint-fix": "npm run lint -- --fix",
-    "test": "jest --runInBand",
-    "ci": "npm run lint && npm test",
+    "test": "NODE_ENV=test jest --runInBand",
+    "ci": "NODE_ENV=test npm run lint && npm test",
```
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply package.json:12]: Definitely is... :P

File: src/lib/nodemailer.js (hunk 1 above)
Comment (imnotteixeira) [reply src/lib/nodemailer.js:28]: Yep, I knew about that (I am using it in the sendMail function), that one must have slipped...

File: src/lib/nodemailer.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    async sendMail(message, options) {
+        if (!this.transporter) {
+            console.warn(`Email service transporter not defined. 
+        Either you forgot to call init(), or MAIL_ADDRESS is not present in .env. Executing no-op.`);
+            return Promise.resolve();
+        }
+
+        const { sendFailureNotification = true } = { ...options }; // To allow undefined options object
```
[the remark this answers is not in the document]
Comment (imnotteixeira) [reply src/lib/nodemailer.js:38]: I think as a developer, if there is a default value, it should be applied whenever it is isn't specified, that includes the empty object. Therefore, I am leaving as is, but I improved the explanation comment and added that test case

File: src/services/application.js (hunk 4 above)
Comment (imnotteixeira) [reply src/services/application.js:10]: Well, it depends... Since the beginning I've always thought of the Reject Reason as an internal thing. It might be really bad PR to have a "bad" reject reason leak in an email by some lousy admin. In the end, it's up to you @andrefmrocha 

The options we have are:
* Use a generic email saying they were rejected (we can give some common causes of rejection such as spam) and if they think they were wrongly rejected to come back to us @ niMail
* Give them the actual reject reason in the email (must be well written every time, so it involves a bigger responsibility when reviewing)

File: src/services/emails/companyApplicationApproval.js (hunk 3 above)
Comment (imnotteixeira) [reply src/services/emails/companyApplicationApproval.js:2]: Will add https://github.com/yads/nodemailer-express-handlebars.

@andrefmrocha if you could review the emails, that would be great. Otherwise just delegate someone instead :upside_down_face:

File: test/end-to-end/application.js (hunk 6 above)
Comment (imnotteixeira) [reply test/end-to-end/application.js:121]: They never failed so far, since they are sent "synchronously", one after the other. But I could use an .includes check on the calls array instead, since the order doens't really matter

File: src/services/application.js (hunk 4 above)
Comment (andrefmrocha) [reply src/services/application.js:10]: I agree that the second option can only be taken if there's a review of such email. 

But perhaps this is not worth the hassle, at least for now. Why don't we go with a generic email for now? Less implementation overhead and overall less risky.

File: src/services/emails/companyApplicationApproval.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    NEW_COMPANY_APPLICATION_COMPANY: (companyName, applicationId) => ({
+        subject: "Your NIJobs Application",
+        text:
+            `We have successfully received your application!
+            \n
+            We will now review your application, and in case you're approved, you will receive another email with further instructions in order to complete your registration. 
+            \n
+            Your Application ID is ${applicationId} and you registered ${companyName}
+            \n
+            If you did not request this or if you need anything else, don't hesitate to contact us!
```
Comment (andrefmrocha) [inline src/services/emails/companyApplicationApproval.js:39]: Even though this does send the contact details in the signature, it's probably best to mention it again here? This seems a very common pattern in these sort of emails. Wdyt?

File: src/services/emails/companyApplicationApproval.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            <br>
+            <p>Sincerely,</p>
+            <p>NIJobs team at NIAEFEUP</p>
+            
+        `,
+    }),
+    APPROVAL_NOTIFICATION: (companyName) => ({
+        subject: "Your NIJobs Application",
+        text:
+            `Glad to have you on board, ${companyName}!
```
Comment (andrefmrocha) [inline src/services/emails/companyApplicationApproval.js:59]: How about: `It is our pleasure to welcome you aboard`?
Not sure if the current one relays that they have been accepted and since this is the first sentence to be seen, even before the email has been opened.

File: src/services/emails/companyApplicationApproval.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+            <br>
+            <p>If you did not request this or if you need anything else, don't hesitate to contact us!</p>
+            <br>
+            <p>Sincerely,</p>
+            <p>NIJobs team at NIAEFEUP</p>
+            
+        `,
+    }),
+    APPROVAL_NOTIFICATION: (companyName) => ({
+        subject: "Your NIJobs Application",
```
Comment (andrefmrocha) [inline src/services/emails/companyApplicationApproval.js:57]: It is common for the subject on an approval message to already mention that the application was accepted in some way, perhaps something simple such as `Welcome to NIJobs!`? Not sure if this is the best one, but something on these terms sounds better

**Review by andrefmrocha — CHANGES_REQUESTED** [review]
Perhaps we can improve these email messages :thinking:

File: src/services/emails/companyApplicationApproval.js (hunk 12 above)
Comment (imnotteixeira) [reply src/services/emails/companyApplicationApproval.js]: Do you mean the subject or the first sentence on the body. If it is the body, they are pretty much the same imo.

Since you commented the subject specifically below, I am confused

Comment (andrefmrocha) [reply src/services/emails/companyApplicationApproval.js]: It was about the first sentence of the body, but with the subject change, I think we can assume that's clear enough.

File: src/services/application.js (hunk 4 above)
Comment (miguelpduarte) [reply src/services/application.js]: +1 on @imnotteixeira's first option. This way we have a consistent "face" to show to the companies. If this was a website aimed at a broader / less "serious" audience, I'd advocate for including the reject reason, as many websites do. However, what you mentioned made me prefer the other option.

As an "intermediate" alternative (for later, as it would be more effort and requires handling things that are not yet implemented) we could have some predefined reject reasons and have a generic email that includes them as simple snippets, based on what the operators chose. If the "other" reason is chosen a text field can be filled for internal consumption and the message would just be the generic part. This might be worth implementing later on when the project is already up and running, so as to provide more feedback to companies.

File: src/email-templates/new_company_application_admins.handlebars [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<h3>A company has requested access to NIJobs</h3>
+<br>
+<p>{{companyName}} <{{email}}> has requested to participate in NIJobs. Below is their motivation:</p>
+<blockquote><i>{{motivation}}</i></blockquote>
```
Comment (miguelpduarte) [inline src/email-templates/new_company_application_admins.handlebars:4]: Any reason to use the `<i>` here? [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/i) seems to suggest it would not make much sense.

File: src/email-templates/rejection_notification.handlebars [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<p>Hello {{companyName}},</p>
+<br>  
+<p>We have reviewed your application for {{companyName}} and decided to reject your application.</p>    
+<p>Some common causes of rejection are spam, abuse and innapropriate motivation.
```
Comment (miguelpduarte) [inline src/email-templates/rejection_notification.handlebars:4]: `s/and/or`

Comment (miguelpduarte) [reply src/email-templates/rejection_notification.handlebars:4]: innapropriate sounds weird as well, we might want to clarify with something like "or a motivation that does not align with the platform's principles" or "our principles". Not sure though, the big boss @andrefmrocha should call this one 😁

File: test/end-to-end/review.js [hunk 16]
```diff
@@ -400,6 +400,25 @@ describe("Company application review endpoint test", () => {
                         expect(res.body.companyName).toBe(pendingApplication.companyName);
                     });
 
+                    test("Should send approval email to company email", async () => {
```
Comment (miguelpduarte) [inline test/end-to-end/review.js:403]: `s/approval/rejection`

**Review by miguelpduarte — CHANGES_REQUESTED** [review]
Some small fixes, almost ready imo

File: src/email-templates/rejection_notification.handlebars (hunk 15 above)
Comment (andrefmrocha) [reply src/email-templates/rejection_notification.handlebars:4]: "or a motivation that does not align with our principles" sounds great in this case.

File: src/email-templates/new_company_application_admins.handlebars (hunk 14 above)
Comment (imnotteixeira) [reply src/email-templates/new_company_application_admins.handlebars:4]: It renders the text as italic. I think it helps to differentiate the text better (the blockquote simply indents it)

File: src/email-templates/rejection_notification.handlebars (hunk 15 above)
Comment (imnotteixeira) [reply src/email-templates/rejection_notification.handlebars:4]: Yep, this was a mere placeholder for review. I didn't really know what to put there... Changing

**Review by andrefmrocha — APPROVED** [review]
Lgtm :shipit:

**Review by miguelpduarte — APPROVED** [review]
